waiter/src/waiter/token.clj

```diff
@@ -157,6 +169,25 @@
                   token-set (->> tokens (map :token) set)]
               (kv/store kv-store owner-key token-set))))))))
 
+;; A token store which is backed by a key-value store.
+(defrecord KeyValueTokenStore [clock synchronize-fn kv-store]
+
+  TokenStore
+
+  (fetch [_ token include-deleted?]
+    (sd/token->token-description kv-store token :include-deleted include-deleted?))
+
+  (store [_ token service-description-template token-metadata]
+    ;; insert the last-update-time if it doesn't exist
+    (let [token-metadata' (merge {"last-update-time" (.getMillis ^DateTime (clock))} token-metadata)]
+      (store-service-description-for-token synchronize-fn kv-store token service-description-template token-metadata')))
+
+  (delete [this token hard-delete?]
+    (let [{:keys [token-metadata]} (fetch this token hard-delete?)]
+      (if-let [owner (get token-metadata "owner")]
+        (delete-service-description-for-token clock synchronize-fn kv-store token owner :hard-delete hard-delete?)
```

Comment:
Inline `delete-service-description-for-token`

Reply:
Same `token-lock` issue.